### ISSUE TITLE
Zimdump crashes on long titles

### ISSUE THREAD

**Issue by cscott** [wikimedia/openzim#2]
Ping @kelson42 @mgautierfr --
From Josh Bullock:
>Zimdump fails and terminates when trying to dump all files within a .zim archive and it comes across a filename that is too long. The filename in my case is:
>
>"A Collection of State-papers, Relative to the First Acknowledgment of the Sovereignty of the United States of America, and the Reception of Their Minister Plenipotentiary, by Their High Mightinesses the States General of the United Netherlands : to Which is Prefixed, the Political Character of John Adams, Ambassador Plenipotentiary from the States of North America, to Their High Mightinesses the States General of the United Provinces of the Netherlands ... Likewise, an Essay on Canon and Feudal Law"

> Asinine, I know. Do you know of any way to fix this, either on my end, or as a hotfix to the code on git?

**Comment by cscott** [comment]
I'm not aware of any hard-coded limits to filename length in the read API, and the [file format](http://www.openzim.org/wiki/ZIM_file_format) uses zero-terminated strings, so it shouldn't have any inherent length limits either.   The given article title is ~500 characters long -- I wonder if you're hitting filesystem [PATH_MAX limits](http://stackoverflow.com/questions/833291/is-there-an-equivalent-to-winapis-max-path-under-linux-unix)... are you running on Windows by any chance?

**Comment by ansabhailte** [comment]
No, I'm on Linux (Mint 17.2 64-bit).

I've tried this on NTFS (external) and EXT4; both fail with the same error.

**Comment by cscott** [comment]
How are you creating the zim file?  Could you post an zim file we can
reproduce the bug with?

**Comment by mgautierfr** [comment]
Both Ext4 and NTFS are limited to 255 characters (bytes for ext4). (https://en.wikipedia.org/wiki/Comparison_of_file_systems)
The zimfile itself is correct (I've created one) but it is just the filesystem not supporting the long filename.

We should truncate the filename in zimdump to be able to dump long titled articles.

**Comment by cscott** [comment]
If we truncate the filename in zimdump, we potentially break any links to
that url.  We probably should (also?) fix this in mwoffliner/zimwriterfs so
that the links are adjusted.

**Comment by mgautierfr** [comment]
No, just truncate in zimdump when we create the file on the filesystem.
The zim itself is valid and zimlib is not buggy, we should not change it.

**Comment by kelson42** [comment]
* This is a limitation of the filesystem, write failure (for filename too long) should be catched by zimdump and one way to deal with could be: Try to write the file with a filename truncated to 255 and write a warning message to STDERR.
* mwoffliner should use the ZIM nodejs binding to fix any potential problem of this nature
* zimwriterfs works the other way, so there is not problem because ZIM allows any filename length anyway.
* zimdump does not rewrite any content anyway... so yes probably the HTML will be broken... but not sure I want to implement url rewritting in zimdump... This is anyway pretty much related to the future of the namespaces in the ZIM format. Not so easy to make a decision for me on that point in particular.

**Comment by ansabhailte** [comment]
Could you add a flag to truncate url's over a certain size? Something like -T for truncate?

**Comment by kelson42** [comment]
@ansabhailte Also possible, even if I'm not 100% convinced this is the best approach.

**Comment by ansabhailte** [comment]
The reason I (my personal use case) want to dump this zim file is to utilize advanced regex searches on the contents. So for me, havinf a few mangled url's isn't an issue. This is why I would like a truncate switch, so that at least I can extract the contents even if sacrificing some metadata.

**Comment by kelson42** [comment]
Names are truncated to 255 characters on the filesystem. Problem fixed :) @ansabhailte if this is OK to you, please close the ticket.

**Comment by ansabhailte** [comment]
Thanks a lot!

**Comment by ansabhailte** [comment]
close

**Comment by kelson42** [comment]
@cscott Looks like none of us can close this ticket on github. Do you?

**Comment by kelson42** [comment]
@cscott Please close the ticket.

**Comment by kelson42** [comment]
Ticket should be closed.
